### PR TITLE
fix: correct representative docs sampling, topic mapping, and image indexing

### LINKED ISSUE #2495: `_extract_representative_docs`: duplicate sampling (`replace=True`) and text-based index mapping select wrong documents
### Describe the bug

Two related bugs in `BERTopic._extract_representative_docs()` (`bertopic/_bertopic.py`) affect the correctness of `representative_docs_`. They are independent but live in the same method, so I'm reporting them together and will fix both in a single PR.

#### Bug 1 — sampling with `replace=True` produces duplicate representative documents

```python
documents_per_topic = (
    documents.drop("Image", axis=1, errors="ignore")
    .groupby("Topic")
    .sample(n=nr_samples, replace=True, random_state=42)
    .drop_duplicates()
)
```

When a topic has fewer than `nr_samples` unique documents (common for small topics), `replace=True` can draw the same document multiple times. Those duplicates are fed into the c-TF-IDF similarity calculation, inflating scores and producing duplicate entries in `representative_docs_`. The trailing `.drop_duplicates()` only removes exact duplicate rows across the whole frame *after* sampling — it does not stop the same document being drawn repeatedly within a single topic.

#### Bug 2 — text-based `in` matching maps documents to the wrong topic

```python
doc_ids = [selected_docs_ids[index] for index, doc in enumerate(selected_docs) if doc in docs]
```

The selected documents are mapped back to their original indices via text membership testing. When the same document text appears in more than one topic (short texts, boilerplate, near-duplicates), `doc in docs` matches the **first** occurrence regardless of which topic the document actually belongs to. This causes:

- documents skipped entirely (their index never matches for the right topic),
- misaligned `doc_ids` ↔ `selected_docs` pairs (wrong similarity scores assigned to wrong documents),
- `representative_docs_` containing documents that don't belong to their assigned topic.

### Reproduction

Bug 1 — duplicate representative docs:

```python
from bertopic import BERTopic
from sklearn.datasets import fetch_20newsgroups

docs = fetch_20newsgroups(subset="all", remove=("headers", "footers", "quotes"))["data"][:500]
topic_model = BERTopic(min_topic_size=5)
topics, _ = topic_model.fit_transform(docs)

for topic_id, docs_list in topic_model.representative_docs_.items():
    if len(docs_list) != len(set(docs_list)):
        print(f"Topic {topic_id}: {len(docs_list)} docs, {len(set(docs_list))} unique")
```

Bug 2 — documents mapped to the wrong topic:

```python
from bertopic import BERTopic

docs = [
    "machine learning is great",   # shared text
    "deep learning neural networks",
    "machine learning is great",   # same text, different topic
    "topic modeling with BERTopic",
]
topic_model = BERTopic(min_topic_size=2)
topics, _ = topic_model.fit_transform(docs)
# representative_docs_ may contain docs mapped to the wrong topic
```

### Proposed fix

- Sample without replacement, capped at the topic's unique-document count (`n=min(nr_samples, len(x))`, `replace=False`), and de-duplicate per `(Topic, Document)` before sampling.
- Track the **positional indices** returned by the similarity/MMR selection and use those to look up `doc_ids`, instead of text membership testing.

### Your contribution

I have both fixes ready in my fork with tests, and will open a PR that closes this issue.

### BERTopic Version

0.17.4

---

_Supersedes #2491 and #2492, which are being closed in favour of this consolidated issue + PR._

### PR DESCRIPTION
**Pull request by pidefrem**
# What does this PR do?

Fixes four related correctness bugs in `BERTopic._extract_representative_docs()` and in the one caller that consumes its `repr_docs_ids` output. They are independent but share the same root area, so they're addressed together.

### Bug 1 — duplicate representative documents from `replace=True` sampling

`_extract_representative_docs` sampled with `sample(n=nr_samples, replace=True)` followed by `.drop_duplicates()`. When a topic has fewer unique documents than `nr_samples`, `replace=True` draws the same document repeatedly; the trailing `.drop_duplicates()` runs *after* sampling and doesn't prevent it. The duplicates inflate the c-TF-IDF similarity calculation and can leave the same document appearing multiple times in `representative_docs_`.

**Fix:** de-duplicate per `(Topic, Document)` before the groupby, and sample without replacement capped at the group size — `sample(n=min(nr_samples, len(group)), replace=False)`.

### Bug 2 — text-based `in` matching maps documents to the wrong topic

Selected documents were mapped back to their original indices with a text membership test:

```python
doc_ids = [selected_docs_ids[index] for index, doc in enumerate(selected_docs) if doc in docs]
```

When the same text appears in more than one topic (short texts, boilerplate, near-duplicates), `doc in docs` matches the first occurrence regardless of topic — producing skipped documents, misaligned `doc_ids` ↔ `selected_docs` pairs, and `representative_docs_` containing documents from the wrong topic.

**Fix:** track the positional indices returned by the similarity/MMR selection and use them to look up `doc_ids`, instead of text membership. Both the MMR and non-MMR branches are corrected.

### Bug 3 — `repr_docs_mappings` keyed by unsorted topic labels

The final mapping was built with `zip(topics.keys(), repr_docs_indices)`, but `repr_docs_indices` is produced by iterating over `labels = sorted(documents_per_topic.Topic.to_list())`. `topics` is the caller-supplied dict, whose key order is not guaranteed to match that sorted order — when it doesn't, every topic in `representative_docs_` gets another topic's documents. The `.groupby("Topic")` upstream already guarantees `labels` is the correct, sorted key sequence.

**Fix:** `zip(labels, repr_docs_indices)`.

### Bug 4 — `VisualRepresentation` treats index labels as positions

`repr_docs_ids` holds DataFrame *index labels*, but `VisualRepresentation._convert_image_to_text` flattened the image column with `documents["Image"].to_numpy().tolist()` and then indexed that plain list with those labels. This is only correct when the documents DataFrame carries a default `RangeIndex`; with any other index (e.g. after filtering, or the sliced frames used in `topics_per_class` / `topics_over_time`) it silently picks the wrong images or raises `IndexError`.

**Fix:** keep `images` as the label-indexed Series and access it with `.loc` (and `.iloc[0]` for the "is this a path?" probe, which is positional by intent).

### Testing

18 tests, all passing:
- `tests/test_dedup_representative_docs.py` — no duplicate representative docs per topic, including the MMR branch and topics smaller than `nr_samples`.
- `tests/test_repr_docs_indexing.py` — representative docs map to the correct topic when identical text spans multiple topics; `repr_docs_ids` are valid index labels on a non-default index; `repr_docs_mappings` stays correct when `topics` keys are not in sorted order.
- `tests/test_representation/test_visual.py` — images are selected by index label, not position, on a non-`RangeIndex` frame.
- `tests/conftest.py` — shared `minimal_topic_model` fixture used by the above.

Defaults are unchanged and the change is backward compatible.

Fixes #2495

## Before submitting
- [ ] This PR fixes a typo or improves the docs (if yes, ignore all other checks!).
- [x] Did you read the [contributor guideline](https://github.com/MaartenGr/BERTopic/blob/master/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes (if applicable)?
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Comment by pidefrem** [comment]
Converted this PR to draft to follow the maintainer instructions: wait for the issue to be acknowledged and a fix path agreed on before opening a PR.
This PR still stays as a basis for further discussion if the issue is acknowledged by the maintainer.
Feel free to close this PR if it is judged not relevant or not in line with the repository's roadmap.
